src/VisualStudio/Core/Def/Implementation/Remote/JsonRpcConnection.cs

```diff
@@ -11,49 +11,129 @@
 using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.PooledObjects;
 using Microsoft.CodeAnalysis.Remote;
 
 namespace Microsoft.VisualStudio.LanguageServices.Remote
 {
-    internal class JsonRpcConnection : RemoteHostClient.Connection
+    internal sealed class JsonRpcConnection : RemoteServiceConnection
     {
         private readonly HostWorkspaceServices _services;
-
-        // communication channel related to service information
         private readonly RemoteEndPoint _serviceEndPoint;
+        private readonly IRemotableDataService _remotableDataService;
+        private IPooledConnectionReclamation? _poolReclamation;
+
+        private bool _disposed;
 
         public JsonRpcConnection(
             HostWorkspaceServices services,
             TraceSource logger,
             object? callbackTarget,
-            Stream serviceStream)
+            Stream serviceStream,
+            IPooledConnectionReclamation? poolReclamation)
         {
+            _remotableDataService = services.GetRequiredService<IRemotableDataService>();
             _services = services;
+
             _serviceEndPoint = new RemoteEndPoint(serviceStream, logger, callbackTarget);
             _serviceEndPoint.UnexpectedExceptionThrown += UnexpectedExceptionThrown;
             _serviceEndPoint.StartListening();
+
+            _poolReclamation = poolReclamation;
+#if DEBUG
+            _creationCallStack = Environment.StackTrace;
+#endif
         }
 
-        private void UnexpectedExceptionThrown(Exception exception)
-            => RemoteHostCrashInfoBar.ShowInfoBar(_services, exception);
+#if DEBUG
+        private readonly string _creationCallStack;
 
-        public override Task InvokeAsync(string targetName, IReadOnlyList<object?> arguments, CancellationToken cancellationToken)
-            => _serviceEndPoint.InvokeAsync(targetName, arguments, cancellationToken);
+        ~JsonRpcConnection()
+        {
+            // this can happen if someone kills OOP. 
+            // when that happen, we don't want to crash VS, so this is debug only check
+            if (!Environment.HasShutdownStarted)
+            {
+                Debug.Assert(false,
+                    $"Unless OOP process (RoslynCodeAnalysisService) is explicitly killed, this should have been disposed!\r\n {_creationCallStack}");
+            }
+        }
+#endif
+        public override void Dispose()
+        {
+            var poolReclamation = Interlocked.Exchange(ref _poolReclamation, null);
+            if (poolReclamation != null)
+            {
+                poolReclamation.Return(this);
+                return;
+            }
 
-        public override Task<T> InvokeAsync<T>(string targetName, IReadOnlyList<object?> arguments, CancellationToken cancellationToken)
-            => _serviceEndPoint.InvokeAsync<T>(targetName, arguments, cancellationToken);
+            if (_disposed)
+            {
+                return;
+            }
```

Comment:
i understand the poolReclamation logic.  i'm not quite sure why we do this _disposed checking/setting after it though.  the flow here isn't quite clear to me.

Reply:
The connection may or may not be pooled. If it is we return it to the pool. If it isn't we dispose it.